File: asml/LC42TrappingRainWater.py
class Solution(object):
    def trap(self, height):
        if not height:
            return 0
        left = [0 for i in range(len(height))]
        right = [0 for i in range(len(height))]
        leftmost = 0
        for i in range(len(height)):
            if height[i] > leftmost:
                leftmost = height[i]
            left[i] = leftmost
        rightmost = 0
        for i in range(len(height)-1, -1, -1):
            if height[i] > rightmost:
                rightmost = height[i]
            right[i] = rightmost
        answer = 0
        for i in range(len(height)):
            if min(left[i], right[i]) > height[i]:
                answer += (min(left[i], right[i])-height[i])
        return answer
        

File: asml/LC42TrappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() == 0) return 0;
        vector<int> left(height.size());
        vector<int> right(height.size());
        int leftmost = 0;
        for (int i = 0; i < height.size(); i++) {
            if (height[i] > leftmost) {
                leftmost = height[i];
            }
            left[i] = leftmost;
        }
        int rightmost = 0;
        for (int i = height.size()-1; i >= 0; i--) {
            if (height[i] > rightmost) {
                rightmost = height[i];
            }
            right[i] = rightmost;
        }
        int answer = 0;
        for (int i = 0; i < height.size(); i++) {
            if (min(right[i], left[i]) > height[i]) {
                answer += min(left[i], right[i])-height[i];
            }
        }
        return answer;
    }
};
